src/main/scala/viper/gobra/translator/implementations/components/OptionImpl.scala

```diff
@@ -0,0 +1,439 @@
+          )()
+        )()
+      )()
+    )(domainName = domainName)
+  }
+
+  /**
+    * {{{
+    * axiom {
+    *   forall o : Option[T] :: { optIsNone(o) } optType(o) == optTypeNone() <==> optIsNone(o)
+    * }
+    * }}}
+    */
+  private lazy val optIsNone_type_axiom : vpr.DomainAxiom = {
+    val oDecl = vpr.LocalVarDecl("o", vpr.DomainType(domainName, Map[vpr.TypeVar, vpr.Type]())(Seq(typeVar)))()
+    val expr = isNone(oDecl.localVar, typeVar)()
+
+    vpr.AnonymousDomainAxiom(
+      vpr.Forall(
+        Seq(oDecl),
+        Seq(vpr.Trigger(Seq(expr))()),
+        vpr.EqCmp(
+          vpr.EqCmp(optType(oDecl.localVar, typeVar)(), optTypeNone(typeVar)())(),
+          expr
+        )()
+      )()
+    )(domainName = domainName)
+  }
+
+  /**
+    * {{{
+    * axiom {
+    *   forall o : Option[T] :: { optIsSome(o) } optType(o) == optTypeSome() <==> optIsSome(o)
+    * }
+    * }}}
+    */
+  private lazy val optIsSome_type_axiom : vpr.DomainAxiom = {
+    val oDecl = vpr.LocalVarDecl("o", vpr.DomainType(domainName, Map[vpr.TypeVar, vpr.Type]())(Seq(typeVar)))()
+    val expr = isSome(oDecl.localVar, typeVar)()
+
+    vpr.AnonymousDomainAxiom(
+      vpr.Forall(
+        Seq(oDecl),
+        Seq(vpr.Trigger(Seq(expr))()),
+        vpr.EqCmp(
+          vpr.EqCmp(optType(oDecl.localVar, typeVar)(), optTypeSome(typeVar)())(),
+          expr
+        )()
+      )()
+    )(domainName = domainName)
+  }
+
+  /**
+    * {{{
+    * domain Option[T] {
+    *
+    *   // Constructors
+    *
+    *   function optSome(e : T) : Option[T]
+    *   function optNone() : Option[T]
+    *
+    *   // Destructor
+    *
+    *   function optGet(o : Option[T]) : T
+    *
+    *   // Constructor Types
+    *
+    *   function optType(o : Option[T]) : Int
+    *   unique function optTypeNone() : Int
+    *   unique function optTypeSome() : Int
+    *   function optIsNone(o : Option[T]) : Bool
+    *   function optIsSome(o : Option[T]) : Bool
+    *
+    *   // Axioms
+    *
+    *   axiom {
+    *     forall e : T :: { optSome(e) } optGet(optSome(e)) == e
+    *   }
+    *
+    *   axiom {
+    *     forall o : Option[T] :: { optGet(o) } optIsSome(o) ==> o == optSome(optGet(o))
+    *   }
+    *
+    *   axiom {
+    *     optType(optNone()) == optTypeNone()
+    *   }
+    *
+    *   axiom {
+    *     forall e : T :: { optSome(e) } optType(optSome(e)) == optTypeSome()
+    *   }
+    *
+    *   axiom {
+    *     forall o : Option[T] :: { optType(o) }
+    *       optType(o) == optTypeNone() && o == optNone() ||
+    *       optType(o) == optTypeSome() && exists e : T :: o == optSome(e)
+    *   }
+    *
+    *   axiom {
+    *     forall o : Option[T] :: { optIsNone(o) } optType(o) == optTypeNone() <==> optIsNone(o)
+    *   }
+    *
+    *   axiom {
+    *     forall o : Option[T] :: { optIsSome(o) } optType(o) == optTypeSome() <==> optIsSome(o)
+    *   }
+    * }
```

Comment:
The domain is indeed quite large now.
How about something like the following:

```
domain Option[T] {

  function some(e: T): Option[T]
  function unit(): Option[T]
  function get(o: Option[T]): T
  function isUnit(o: Option[T]): Bool

  axiom {
    forall e: T :: { some(e) } get(some(e)) == e && isUnit(some(e)) == false
  }

  axiom {
    forall o: Option[T] :: { get(o) } isUnit(o) == false ==> o == some(get(o))
  }

  axiom {
    isUnit((unit(): Option[T])) == true
  }

  axiom {
    forall o: Option[T] :: { isUnit(o) } o == unit() || exists e: T :: {some(e)} o == some(e)
  }
}
```

With some tests:
```
domain Dummy[T] {
  function dummy(x: T): Bool
  axiom {
    forall x: T :: { dummy(x) } dummy(x)
  }
}


method test1() {
  var x: Option[Int] := some(4)
  assert x != unit()
  assert get(x) == 4
}

method test2(x: Option[Int]) {
  if (x != unit()) {
    assert dummy(get(x)) // hard to avoid
    assert exists e: Int :: some(e) == x
  }
}
```